src/onepass.rs

```diff
@@ -0,0 +1,1187 @@
+    /// We don't support multiple regex at once.
+    RegexSetUnsupported,
+    /// Hints that destructuring should not be exhaustive.
+    ///
+    /// This enum may grow additional variants, so this makes sure clients
+    /// don't count on exhaustive matching. (Otherwise, adding a new variant
+    /// could break existing code.)
+    #[doc(hidden)]
+    __Nonexhaustive,
+}
+
+impl From<Error> for OnePassError {
+    fn from(e: Error) -> Self {
+        OnePassError::CompileError(e)
+    }
+}
+
+impl OnePassCompiler {
+    /// Create a new OnePassCompiler for a given Hir.
+    /// Collect some metadata from the compiled program.
+    pub fn new(
+        es: &[Hir],
+        options: &RegexOptions,
+        only_utf8: bool,
+    ) -> Result<Self, OnePassError> {
+        if es.len() != 1 {
+            return Err(OnePassError::RegexSetUnsupported);
+        }
+
+        if ! es[0].is_onepass() {
+            return Err(OnePassError::HasNondeterminism);
+        }
+
+        let prog = Compiler::new()
+                         .size_limit(options.size_limit)
+                         .bytes(true)
+                         .only_utf8(only_utf8)
+                         .compile(es)?;
+
+        let num_byte_classes = (prog.byte_classes[255] as usize) + 2;
+
+        // We don't allow STATE_MAX to actually be used so that STATE_POISON
+        // remains a valid poison value.
+        let max_table_size = (STATE_MAX - 1) as usize;
+        let mut table_size: usize = 0;
+        for inst in prog.iter() {
+            table_size += num_byte_classes;
+            match inst {
+                &Inst::EmptyLook(_) | &Inst::Save(_) =>
+                    table_size += num_byte_classes,
+                _ => {}
+            }
+            if table_size > max_table_size {
+                return Err(OnePassError::TooBig);
+            }
+        }
+
+        trace!("new compiler for:\n{:?}", prog);
+        Ok(OnePassCompiler {
+            table: vec![],
+            num_byte_classes: num_byte_classes,
+            only_utf8: only_utf8,
+
+            transitions: {
+                let mut x = Vec::new();
+                for _ in 0..prog.len() {
+                    x.push(None);
+                }
+                x
+            },
+            accepting_states: vec![false; prog.len()],
+            prog: prog,
+        })
+    }
+
+    /// Attempt to compile the regex to a OnePass DFA
+    pub fn compile(mut self) -> Result<OnePass, OnePassError> {
+        // A DAG of forwarding relationships indicating when
+        // a state needs to be forwarded to an Action state
+        // once that Action state has been fully constructed.
+        let mut forwards = Forwards::new();
+
+        // Compute the prioritized transition tables for all of the
+        // instructions which get states.
+        let mut state_edge = vec![0];
+        while let Some(i) = state_edge.pop() {
+            state_edge.extend(self.inst_trans(i, &mut forwards)?);
```

Comment:
Extending `state_edge` with a vector here will cause an allocation per instruction. I should just pass a mutable reference to `state_edge` into `inst_trans` and mutate it directly instead.